### ISSUE TITLE
please remove options.filename requirement for includes

### ISSUE THREAD

**Issue by jasonswearingen** [tj/ejs#137]
simply commenting lib/ejs.js line 158 out and this works:

```
if (!filename) throw new Error('filename option is required for includes');
```

of course it would be nice to throw this error if the next line (resolveInclude) fails, but if it doesn't fail, no reason to force people to add this options.filename verbosity.

ps: i don't take credit for this hack solution,  see @Ryan McCaffery here: http://stackoverflow.com/questions/13537032/node-js-ejs-include-error

**Comment by ORESoftware** [comment]
I just ran into this one, StackOverflow doesn't seem to solve it short of commenting out the source

**Comment by ForbesLindesay** [comment]
The correct way to solve it is to pass in `opyions.filename`, which should be the filename of the EJS file you are compiling/rendering.

**Comment by juuxstar** [comment]
I disagree.  Sometimes there is no file for the template (it could be dynamically composed, or maybe pulled from a DB, or any other non-file source)

It looks to me like the reason for the `options.filename` parameter is to solve the problem of resolving relatively referenced includes.  But IMO this should be addressed as the problem that it is.  Meaning, have a way for the user to specify how to resolve relative filenames (for instance with the callback I suggested a while ago)

**Comment by mde** [comment]
This version of EJS (v1) is no longer active or supported. The current version on NPM is v2, actively maintained here: https://github.com/mde/ejs

**Comment by ORESoftware** [comment]
THANKS MDE!! can you install your version with Bower? Because when I installed with Bower it probably gave me the old version.

**Comment by ORESoftware** [comment]
ummm, I was looking at the ejs source, it's trying to require('fs') and stuff, and I get an error saying readFileSync is not a function. I am familiar with node.js, but I am trying to use EJS on the front-end right now. Why is it trying to do an fs.readFileSync? Maybe it's because I pass in options...I know it's a lot to ask but maybe, just maybe, there should be an EJS for front-end and an EJS for back-end.

**Comment by ORESoftware** [comment]
I think the problem I am encountering is this:

(1) I want to use includes
(2) To use includes, I would have to pass in options (Error: `include` requires the 'filename' option.)
(3) When I pass in options the library seems try to try to use fs and fs.readFileSync (fails on front-end)

so I get an error (fs.readFileSync is not a function)

is this correct?

**Comment by mde** [comment]
EJS v2 is bundled for browser use using Browserify. You can download the latest release (https://github.com/mde/ejs/releases/latest), then use the ejs.js or ejs.min.js file.

If you're using EJS on the browser side, I'm not sure what the `include` function would be expected to do. It's meant to be used to include other files, and the browser doesn't have access to the filesystem.

**Comment by ORESoftware** [comment]
yeah I see

(a) include could do AJAX (not a good idea)
(b) include could look for files adjacent to the current one. I am using RequireJS which bundles the templates all in one file.

there must be a way! How do I do includes on the front-end? I don't get it.

**Comment by ORESoftware** [comment]
also, one more question

if we have to pass a filename option, how do we include multiple templates?

is there a filenames (plural) option?

**Comment by ForbesLindesay** [comment]
Compile on the server using node.js then use the compiled code on the client side.  Includes will then work fine.

**Comment by ORESoftware** [comment]
thanks FL  - so on the server, when I compile a template file #1 I imagine that I put the compiled code into another file #2 and then in the template with the include statement I read from file #2? is there an example of this anywhere?

However, before I make that drastic change:

(a) is there a good way to get EJS include behavior without using includes?
(b) on the client isn't there simply a good way to stuff a compiled template into another template beyond just using local variables?
